File: leetcode/4sum2/optimal.py
class Solution:
    def fourSumCount(self, A: List[int], B: List[int], C: List[int], D: List[int]) -> int:
        if len(A) == 0:
            return 0
        
        m1 = {}
        for a in A:
            if a in m1:
                m1[a] += 1
            else:
                m1[a] = 1
        m2 = {}
        for a, v in m1.items():
            for b in B:
                ab = a + b
                if ab in m2:
                    m2[ab] += v
                else:
                    m2[ab] = v
        
        m3 = {}
        for c in C:
            if c in m3:
                m3[c] += 1
            else:
                m3[c] = 1
        
        res = 0
        for c, v in m3.items():
            for d in D:
                cd = - c - d
                if cd in m2:
                    res += m2[cd] * v
                
        return res

File: leetcode/getAllElements/optimal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    void rec(TreeNode* root, vector<int>& v){
        if (!root) return;
        rec(root->left, v);
        v.push_back(root->val);
        rec(root->right, v);
    }
    
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {
        vector<int> res, v1, v2;
        rec(root1, v1);
        rec(root2, v2);
        int i = 0, j = 0;
        while(i<v1.size() && j<v2.size()){
            if (v1[i]<=v2[j]) res.push_back(v1[i++]);
            else res.push_back(v2[j++]);
        }
        
        while(i<v1.size()) res.push_back(v1[i++]);
        while(j<v2.size()) res.push_back(v2[j++]);
        return res;
    }
};